### ISSUE TITLE
AxisLegend

### ISSUE THREAD

**Issue by KronosTheLate** [JuliaPlots/Makie.jl#826]
Currently, if I want to create a legend inside some axis, I would have to save the output of e.g. `plot(xs, ys)` into a variable (still getting used to this, coming from Plots.jl). Then I would have to create a legend object, which I need to place in the previously created figure-object, and pass the plot-objects to the Legend-object. Addidionally, I have to say that `tellheight` and `tellwidth` should be false. That is fine for functionality, but strikes me as poorly optimized for convenience. I for one find it very tedeous when I want a simple legend that does not add width and shrink my data.

I am proposing the following as a solution:
Define a new object AxisLegend. This object is associated with some axis as opposed to the figure - it's first argument is an axis-object which it gets placed inside, as opposed to a figure-object with the need for explicit placement. It never tells it's heigh or width, and you don't have to create plot objects for passing to it - it only needs a vector of strings of equal length to the number of plots in the Axis. I am hoping that the plot objects are ordered by their order of addition to the Axis object, and that this order would be used in matching strings to Plot-objects. This functionality would mimic the syntax that I for one know and love from Plots.jl. The AxisLegend would also need a location kwarg, preferably with some auto setting as default that would simply try to not obstruct data. Optimally, this object could be created inside a `plot`/`scatter`/`lines` command, which would pass it directly to the axes associated with the plot-series. This would remove the need to define figure, axis and plot object and seperatly add a legend.

In other words, I am imagining that this would allow a simple one-liner to create a simple plot with an embedded legend, without having 3 object to define and pass around, without manual placement of the legend, and without having to explicitly define the legend not to tell its height and width.

This does not add functionality, so it should not be a priority. But I would love for the possibility of quicking visualizing stuff, without needing multiple wordy lines of code to make the visualization nice.

An alternative name for this object could be EmbeddedLegend, but I think that AxisLegend is shorter and clearer.

**Comment by jkrumbiegel** [comment]
> That is fine for functionality, but strikes me as poorly optimized for convenience

I agree completely, it's just what happens when you build a more general system instead of a single optimized use case. This problem is not just related to the Legend, for example it's currently annoying to create a horizontal Colorbar, just because multiple defaults which are meant for the vertical default have to be changed.

I'm still thinking about what mechanism is the easiest / clearest to use. Defining a new object is not so optimal, I do want a Legend to stay a Legend always. Maybe a simple function that predefines those defaults which we need for the in-axis case is better.

**Comment by KronosTheLate** [comment]
I agree, creating something like a `axislegend` function is a more extendable and less dramatic alternative to defining a new type.

If however it is a general problem in the very-extendable API of Makie that expressing simple and normal things can be long-winded, then maybie a new issue should be created that more accuratly describes the general problem? Then I can post an example of how I currently can create an axislegend, vs how I would prefer to do it, with code examples instead of multiple paragraphs. And then for example you can chime in with an example of what the problem looks like for horizontal colorbars. And solution can be proposed to each problem, and possible overlaps or contradictions in the solutions can be spotted. Maybe then it is easier to get an overview of the full problem, which can make it easier to arrive at a complete and unified solution to the general problem.

Do you agree that that is better, and that I should close this issue to the advantage of the more general one?

**Comment by jkrumbiegel** [comment]
I've added an `axislegend` function now, should any really good idea come up later how to generalize presets for different layoutables we can think about it then. For now I thought it's more important to make this functionality easily available. http://makie.juliaplots.org/stable/makielayout/llegend.html#Legend-Inside-An-Axis
